src/phone-input/stateful-phone-input-container.js

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright (c) 2018-2019 Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+import React from 'react';
+import {countries, STATE_CHANGE_TYPE} from './constants.js';
+
+import type {
+  StatefulPhoneInputContainerPropsT,
+  StateT,
+  StateReducerT,
+  StateChangeT,
+  StateChangePayloadT,
+  CountryChangeEventT,
+} from './types.js';
+
+const getCountry = iso => {
+  return countries.find(({id}) => id === iso) || {dialCode: ''};
+};
+
+const defaultStateReducer: StateReducerT = (type, nextState) => nextState;
+
+export default class StatefulPhoneInputContainer extends React.Component<
+  StatefulPhoneInputContainerPropsT,
+  StateT,
+> {
+  static defaultProps = {
+    initialState: {
+      inputValue: `+${getCountry('US').dialCode} `,
+      countryValue: getCountry('US'),
+    },
+    overrides: {
+      Input: {},
+      CountrySelect: {},
+      CountrySelectDropdown: {},
+      CountrySelectDropdownListItem: {},
+      CountrySelectDropdownFlagColumn: {},
+      CountrySelectDropdownNameColumn: {},
+      CountrySelectDropdownDialcodeColumn: {},
+    },
+    onInputChange: () => {},
+    onCountryChange: () => {},
+    stateReducer: defaultStateReducer,
+  };
+
+  state = {...this.props.initialState};
+
+  internalSetState = (type: StateChangeT, payload: StateChangePayloadT) => {
+    const nextState: StateT = this.internalStateReducer(type, payload);
+    this.setState(this.props.stateReducer(type, nextState));
+  };
+
+  internalStateReducer = (type: StateChangeT, payload: StateChangePayloadT) => {
+    switch (type) {
+      case STATE_CHANGE_TYPE.inputValueChange:
+        if (typeof payload === 'string') {
+          return {...this.state, inputValue: payload};
```

Comment:
We should be able to pass he only the value that is changed, so no need to spread the current state.